keps/sig-node/127-user-namespaces/README.md

```diff
@@ -803,6 +839,11 @@ logs or events for this purpose.
 
 ###### How can someone using this feature know that it is working for their instance?
 
+Check if any pod has the pod.spec.HostUsers field set to false and is on RUNNING state on a node
+that meets all the requirements.
+
```

Comment:
@wojtek-t Why do we need that? I'm honestly not sure I understand, let me explain why:

* We are not doing enforcement of any kind based on the feature gate `UserNamespacesSupport`, so there are no security concerns for requesting this and then not being enabled.
* AFAIK we don't have that metric in the kubelet for any other namespace (nor share the PID namespace, that was added in like 1.17, quite "recently"). **Why is this namespace special?** I honestly don't see it.
* Even if we do want a metric, isn't the container runtime a better place for this metric?
* The CRI runtime doesn't really know if this was honored or not. For runc-compatible runtimes, _maybe_ we can manage to get it (using the features subcommand if idmap is supported, then it must error-out if its not honored). But there is no real feedback from the OCI runtime on what was honored or not, let alone non-OCI runtimes. 

So, I really want to push back a little here and understand why we need this and why we need this in the kubelet metrics.

Reply:
> Why is this namespace special? 

It's not special, but past mistakes don't justify future mistakes.

I think we have a real plumbing problem, though.  If you have a CRI which does not understand this new field, it can't report that it doesn't understand it...because it doesn't know what it doesn't know.  If we don't have a way to interrogate the runtime, how can we get this info (except to try it and look at the results)?

IOW, ideally it would be something like:

kubelet: Dear CRI, do you support userns?
runtime: No
kubelet: Dear PodStatus, this pod asked for userns but did not get it

But if there isn't a way to ask the question, it's a lot more complicated

Reply:
We can use RuntimeConfig CRI API for this - https://github.com/kubernetes/kubernetes/commit/0f05a92ce68ec8ae5b65aa4db746300aba5121ca

Reply:
Thanks @thockin for making this explicit - this is exactly my concern.
I agree that this feature isn't special, but exactly - the fact that we did something bad wrong in the past doesn't mean we can repeat mistakes.

If as a user I'm asking for X (in this case user namespace) and my pod gets to RUNNING state, I really expect it to be running in user namespace. If it's not (and it silently dropped that intent), the actual impact doesn't matter that much - it's just bad because we didn't something else than the user asked.
I would expect the pod to be rejected, and if not, at the very very least see a easy to digest for the operator signal that shows that (e.g. a metric on kubelet and a status on the pod)....

Reply:
would you be ok with something along the lines of the following check?

```diff
diff --git a/pkg/kubelet/kuberuntime/util/util.go b/pkg/kubelet/kuberuntime/util/util.go
index af61e17567e..a9f0a33c5d5 100644
--- a/pkg/kubelet/kuberuntime/util/util.go
+++ b/pkg/kubelet/kuberuntime/util/util.go
@@ -18,6 +18,7 @@ package util
 
 import (
        v1 "k8s.io/api/core/v1"
+       internalapi "k8s.io/cri-api/pkg/apis"
        runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
        "k8s.io/klog/v2"
        kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
@@ -99,11 +100,20 @@ func PidNamespaceForPod(pod *v1.Pod) runtimeapi.NamespaceMode {
 
 // namespacesForPod returns the runtimeapi.NamespaceOption for a given pod.
 // An empty or nil pod can be used to get the namespace defaults for v1.Pod.
-func NamespacesForPod(pod *v1.Pod, runtimeHelper kubecontainer.RuntimeHelper) (*runtimeapi.NamespaceOption, error) {
+func NamespacesForPod(ctx context.Context, pod *v1.Pod, runtimeHelper kubecontainer.RuntimeHelper, runtimeService internalapi.RuntimeService) (*runtimeapi.NamespaceOption, error) {
        userNs, err := runtimeHelper.GetOrCreateUserNamespaceMappings(pod)
        if err != nil {
                return nil, err
        }
+       if userNs != nil || userNs.Mode != runtimeapi.NamespaceMode_NODE {
+               cfg, err := runtimeService.RuntimeConfig(ctx.Background())
+               if err != nil {
+                       return nil, err
+               }
+               if !runtime.Linux.UserNamespaceSupported {
+                       return nil, fmt.Errorf("The runtime doesn't support user namespaces")
+               }
+       }
```

Reply:
Note: this modulo the feature gate checking :)

Reply:
instead of using a single bool for the RuntimeConfig, we can use the same RuntimeHandlerFeatures mechanism described for KEP-3857 (Recursive Read-Only mounts): https://github.com/kubernetes/enhancements/pull/3858/files#diff-057b8627f24bc6a0742b51b4fea113938a3440eafe255edecc834f3301c008fcR478 so we can specify the UserNamespace feature support for each separate handler

Reply:
Yes, that approach will work for both the cases and also handle runtime class specific features.

Reply:
> Just to be clear: this will only create the pod if the CRI runtime knows about userns, and return the error otherwise.

Yes - that's exactly what I would like to have.
Can you please update the PRR (and probably mention it in the core of the KEP too) to reflect that?

Reply:
Sure! Now that we agreed on a direction, we will update the KEP :)

Reply:
We updated the kep @wojtek-t , PTAL!

Reply:
Thank you - resolving and will add more targetted comments.